File: main/src/main/java/io/tripled/social/client/SocialNetworkApplication.java
package io.tripled.social.client;

import io.tripled.social.client.application.UseCaseFactory;
import io.tripled.social.client.domain.DateTimeProvider;
import io.tripled.social.client.domain.SocialNetworkRepository;
import io.tripled.social.client.presentation.Input;
import io.tripled.social.client.presentation.Output;
import io.tripled.social.client.presentation.ReadEvalPrintLoop;
import io.tripled.social.client.presentation.console.ConsoleInput;
import io.tripled.social.client.presentation.console.ConsoleOutput;
import io.tripled.social.client.presentation.controller.FollowController;
import io.tripled.social.client.presentation.controller.PostController;
import io.tripled.social.client.presentation.controller.ReadController;
import io.tripled.social.client.presentation.controller.WallController;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Optional;
import java.util.ServiceLoader;

public class SocialNetworkApplication {

    public static void main(String[] args) {
        ReadEvalPrintLoop repl = createRepl();

        new Thread(repl).start();
    }

    private static ReadEvalPrintLoop createRepl() {
        Input input = new ConsoleInput(new BufferedReader(new InputStreamReader(System.in)));
        Output output = new ConsoleOutput();

        final DateTimeProvider dateTimeProvider = loadSystemDateProvider();
        final SocialNetworkRepository socialNetworkRepository = loadSocialNetworkRepository();
        final UseCaseFactory useCaseFactory = new UseCaseFactory(socialNetworkRepository);
        return new ReadEvalPrintLoop(input, output,
                new PostController(useCaseFactory.createPostMessageUseCase(dateTimeProvider)),
                new ReadController(useCaseFactory.createReadMessagesUseCase()),
                new WallController(useCaseFactory.createReadWallUseCase()),
                new FollowController(useCaseFactory.createFollowUserUseCase()));
    }

    private static DateTimeProvider loadSystemDateProvider() {
        final ServiceLoader<DateTimeProvider> repositories = ServiceLoader.load(DateTimeProvider.class);
        final Optional<DateTimeProvider> first = repositories.findFirst();
        return first.orElseThrow(RuntimeException::new);
    }

    private static SocialNetworkRepository loadSocialNetworkRepository() {
        final ServiceLoader<SocialNetworkRepository> repositories = ServiceLoader.load(SocialNetworkRepository.class);
        final Optional<SocialNetworkRepository> first = repositories.findFirst();
        return first.orElseThrow(RuntimeException::new);
    }
}

File: src/main/kotlin/io/tripled/social/client/SocialNetworkApplication.kt
package io.tripled.social.client

import io.tripled.social.client.application.usecase.FollowUserUseCase
import io.tripled.social.client.application.usecase.PostMessageUseCase
import io.tripled.social.client.application.usecase.ReadMessagesUseCase
import io.tripled.social.client.application.usecase.ReadWallUseCase
import io.tripled.social.client.domain.DateTimeProvider
import io.tripled.social.client.infrastructure.cli.Input
import io.tripled.social.client.infrastructure.cli.Output
import io.tripled.social.client.infrastructure.cli.ReadEvalPrintLoop
import io.tripled.social.client.infrastructure.cli.console.ConsoleInput
import io.tripled.social.client.infrastructure.cli.console.ConsoleOutput
import io.tripled.social.client.infrastructure.cli.controller.FollowController
import io.tripled.social.client.infrastructure.cli.controller.PostController
import io.tripled.social.client.infrastructure.cli.controller.ReadController
import io.tripled.social.client.infrastructure.cli.controller.WallController
import io.tripled.social.client.infrastructure.inmemory.InMemorySocialNetworkRepository
import io.tripled.social.client.infrastructure.time.SystemDateTimeProvider
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val repl = createRepl()
    Thread(repl).start()
}

private fun createRepl(): ReadEvalPrintLoop {
    val input: Input = ConsoleInput(BufferedReader(InputStreamReader(System.`in`)))
    val output: Output = ConsoleOutput()

    val dateTimeProvider: DateTimeProvider = SystemDateTimeProvider()
    val socialNetworkRepository = InMemorySocialNetworkRepository(dateTimeProvider)

    return ReadEvalPrintLoop(
        input,
        output,
        PostController(PostMessageUseCase(socialNetworkRepository, dateTimeProvider)),
        ReadController(ReadMessagesUseCase(socialNetworkRepository)),
        WallController(ReadWallUseCase(socialNetworkRepository)),
        FollowController(FollowUserUseCase(socialNetworkRepository))
    )
}